Issue #308: Preview for email templates

jlwt90 opened the pull request:
Hello, this is the PR for #308 .

I modified the behaviour of the first email template: *Placed Order*. 
![screen shot 2017-03-25 at 4 35 31 pm](https://cloud.githubusercontent.com/assets/25281597/24320516/3bf5cf62-117a-11e7-971d-d3bd172c56e7.png)

There are two tabs at the bottom, *edit* and *preview*.
When you click on the *preview* tab, an ajax request will be fired to `/control/event/**/***/settings/email/preview` endpoint.

![screen shot 2017-03-25 at 4 35 57 pm](https://cloud.githubusercontent.com/assets/25281597/24320542/8c31b586-117a-11e7-9423-1ca3cde9f1ec.png)

I haven't implemented the backend logic at this moment (I'll add more commits later). 
The dummy response I am currently using is:
```
{
            'item': 'order_placed',
            'msgs': {
                'mail_text_order_placed_0': 'English version!',
                'mail_text_order_placed_1': 'Deutsch version!'
            }
        }
```
What do you think about this UI? :D
PS: Please also help check whether I can improve my code too. I don't have much front-end experience. Comments are welcome :)

*Task:*
- [x] Draft of UI change
- [x] UI Modification based on comments
- [x] Remove dummy response in backend view
- [x] Work on the remaining templates
- [x] Add tests
- [x] Fix locale index issue
- [x] Improve dummy data
- [x] Display date based on locale

raphaelm commented on src/pretix/control/templates/pretixcontrol/event/mail.html:
```diff
@@ -21,9 +21,29 @@ <h4 class="panel-title">
                         <div class="panel-body">
-                            {% bootstrap_field form.mail_text_order_placed layout="horizontal" %}
+                            <div>
+                                <label class="col-md-3 control-label">Text</label>
```
``{% trans "Text" %}`` please, just so we don't forget later :)

raphaelm commented on src/pretix/static/pretixcontrol/js/ui/mail.js:
```diff
@@ -0,0 +1,83 @@
+function preview_task_callback(data, jqXHR, status) {
+    "use strict";
+    if (data.item){
```
Just a formatting nitpick, but a space between ``)`` and ``{``

jlwt90 commented:
@raphaelm Thanks!

>but if they send the form to save, the textareas will be sent twice to the server, one time with the changed content and one time with the unchanged content which might lead to unexpected behaviour

I also have the same concern when I wrote that part. My original idea is to keep the preview textarea with exactly the same style so that the transition from `edit` mode to `preview` mode looks really smooth. I was struggling whether I should include the dynamic block `{% bootstrap_field %}` twice so as to prevent duplicated code in `mail.html`. The code in current version is cleaner but kinda risky (I totally agree with you!). 

>Don't use {% bootstrap_field but construct the preview fields manually in the template

Let's pick this one. Since preview feature is just for improving user experience, I don't think it's appropriate to change backend form implementation (option 2).

raphaelm commented:
> My original idea is to keep the preview textarea with exactly the same style so that the transition from edit mode to preview mode looks really smooth

I appreciate this thought! Although I think that it wouldn't be that much of a problem if the preview looked different than the editor. But I think it should be possible to replicate the textareas nevertheless.

jlwt90 commented:
>I appreciate this thought! Although I think that it wouldn't be that much of a problem if the preview looked different than the editor

After I replaced `textarea` with `p`, I feel the same. :)
I think this one looks better?
Here is the screenshot:
![screen shot 2017-03-27 at 1 07 59 am](https://cloud.githubusercontent.com/assets/25281597/24332932/29cf67b0-128a-11e7-9997-17391d78450b.png)

Regarding the corresponding view, could you give me some directions?
I read some code in some other preview feature, e.g. `TicketSettingsPreview`. Apparently I should create a dummy event and order with transaction rollback, then call a method for generating mail content (`src/pretix/base/services/mail.py`?). 
Thanks!

raphaelm commented:
> After I replaced textarea with p, I feel the same. :)
> I think this one looks better?

Yes, looks good! I believe in the sendmail plugin we even use ``<pre>`` for the previews, but I'm not sure if that would look pretty here. (You can activate the sendmail plugin at Settings > Plugins, which will create a new item in the menu on the left featuring an email form with preview).

> I read some code in some other preview feature, e.g. TicketSettingsPreview. Apparently I should create a dummy event and order with transaction rollback, then call a method for generating mail content (src/pretix/base/services/mail.py?).

Nooooo! That is an ugly, ugly hack :see_no_evil: We need it for tickets, because tickets might be generated by plugins that know nothing about previews. We don't have this problem here, so we don't need to replicate this :D

Instead, it's pretty simple. Just take the email text sent by the user, and call ``.format()`` on it with useful sample values and… done! The available placeholders are listed in the help texts of the form and you could have a look at Flavia's work in PR #279 which did a similar thing to the sendmail plugin.

jlwt90 commented:
>Yes, looks good! I believe in the sendmail plugin we even use \<pre\> for the previews, but I'm not sure if that would look pretty here. (You can activate the sendmail plugin at Settings > Plugins, which will create a new item in the menu on the left featuring an email form with preview).

Thanks! I'll take a look!

>Nooooo! That is an ugly, ugly hack 🙈 We need it for tickets, because tickets might be generated by plugins that know nothing about previews. We don't have this problem here, so we don't need to replicate this :D

Also good to know :D I was thinking why I need to play with database just for rendering the content! Indeed I thought there is an existing method in services layer which simply generates email content based on given event & order (and template for sure) without sending a mail. If this preview feature uses the same logic as the real one, we can guarantee that the preview will be exactly the same as the real mail. We also need to maintain both `MailPreviewView` and mail sending logic if we add/remove new fields. Is my understanding correct?

Thanks again for your detailed explanation!

raphaelm commented:
> I thought there is an existing method in services layer which simply generates email content based on given event & order (and template for sure) without sending a mail. If this preview feature uses the same logic as the real one, we can guarantee that the preview will be exactly the same as the real mail. We also need to maintain both MailPreviewView and mail sending logic if we add/remove new fields. Is my understanding correct?

Sadly: yes, we have this logic twice then. The ``mail`` module in the services layer just takes a template and a dictionary of things and combines them. Maintaining that logic twice is no problem, as its only a ``.format`` call. The interesting logic resides within the *calls* to ``mail``. It would be interesting to refactor that in a way that prevents us from keeping the logic twice -- but maybe that's a target for your next pull request? 😉

jlwt90 commented:
Hello @raphaelm , this commit contains
- \<pre\> preview block
- skeleton of backend preview logic

Please take a look if you have time :D
*Edit:*
![screen shot 2017-03-29 at 1 09 40 am](https://cloud.githubusercontent.com/assets/25281597/24415460/87ec0400-141c-11e7-8651-f66a2065f195.png)

*Preview:*
![screen shot 2017-03-29 at 1 09 31 am](https://cloud.githubusercontent.com/assets/25281597/24415471/8faf9af8-141c-11e7-85d2-4c6cb2a201a3.png)

raphaelm commented on src/pretix/control/templates/pretixcontrol/event/mail.html:
```diff
@@ -23,7 +23,29 @@ <h4 class="panel-title">
+                                        </li>
+                                    </ul>
+                                </div>
+                            </div>
```
Just a thought: We will need all of this boilerplate for every single mail template. I think once the basic setup is clear, we should put this into a seperate template (i.e. ``mail_settings_fragment.html``) that we [{% include %}](https://docs.djangoproject.com/en/1.10/ref/templates/builtins/#include).

jlwt90 replied:
Nice idea! Had a quick look at django template engine doc.
`{% include "name_snippet.html" with person="Jane" greeting="Hello" %}`
I think we can probably use this technique to render all the previews with parameter `preview_item`, e.g. `mail_text_order_placed, mail_text_order_paid`.

Or even more aggressive approach, we can use `{% include %}` to avoid writing duplicated `panel` block, i.e.
```
<div class="panel panel-default">
    <div class="panel-heading">
        <h4 class="panel-title">
            <a data-toggle="collapse" href="#order_paid">
                <strong>{% trans "Paid order" %}</strong>
            </a>
        </h4>
    </div>
    <div id="order_paid" class="panel-collapse collapse">
       <div class="panel-body">
           {% bootstrap_field form.mail_text_order_paid layout="horizontal" %}
       </div>
    </div>
</div>
```
The only differences between two mail panels are:
- Panel title, i.e. `{% trans "Paid order" %}`
- Panel id
- form element, i.e. `form.mail_text_order_paid`

What's your opinion? (I'll give a try tomorrow 👍 )

raphaelm replied:
Please, go for the "aggressive" option with less repetition!

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -399,6 +403,56 @@ def post(self, request, *args, **kwargs):
+    # return the origin text if key is missing in dict
+    class SafeDict(dict):
+        def __missing__(self, key):
+            return '{' + key + '}'
```
Nice approach! :+1:

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -399,6 +403,56 @@ def post(self, request, *args, **kwargs):
+        if preview_item not in self.items:
+            raise HttpResponseBadRequest(_('invalid item'))
+
+        regex = r"^" + re.escape(preview_item) + r"_\d{1}$"
```
I don't think it's a good idea to hardcode that we'll never have more than 10 languages in an event (although it is unlikely, I agree).

jlwt90 replied:
:D agree, I'll accept all valid `preview_item` followed by any number.

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -399,6 +403,56 @@ def post(self, request, *args, **kwargs):
+    def dummy_data(self):
+        return {
+            'mail_text_order_placed': {
+                'event': str(self.request.event.name),
```
If you not cast this to string here, it should still work fine -- and will make things easier if we want the localized versions later.

jlwt90 replied:
Yes, you are right! Apparently only this line failed if `str` is not used.
`'invoice_name': _('Invoice for %s.') % {str(self.request.event.name)},`

raphaelm replied:
You can fix that by using new-style formats here as well:

```
'invoice_name': _('Invoice for {}.').format(self.request.event.name)
```

They can deal with the internationalized strings. In the long run, we eventually want to use ``.format()`` everywhere.

raphaelm commented on src/pretix/control/templates/pretixcontrol/event/mail.html:
```diff
@@ -92,7 +114,7 @@ <h4 class="panel-title">
                         </h4>
                     </div>
-                    <div id="order_expirew" class="panel-collapse collapse">
+                    <div id="order_expire" class="panel-collapse collapse">
```
``order_expirew`` was supposed to be a (apparently bad) abbreviation for "expire *w*arning". If you change it, change the reference in the ``href=`` argument above as well.

jlwt90 replied:
oops! I thought it's just a typo 😭 
Thanks for your reminder, I forgot to change the ref. as well. I'll just change it back!

jlwt90 commented:
@raphaelm Hello, could you review again? This version contains `mail_preview_fragment.html` and custom filter `mail_settings_preview.py`. 

1) The fragment template consists of a collapsable panel with a set of preview panels.
When we include the template, the following key arguments should be provided:
- **pid**: panel ID
- **title**: name display on panel header
- **items**: item names delimited by `,`, e.g. `mail_text_resend_link,mail_text_resend_all_links` for **Resend link** panel

2) Added locale in dummy data generation
As discussed, I assume that `mail_text_resend_link_0` is mapped to the default locale, and the indices are the same as the template, e.g. 0 for en, 1 for de

raphaelm commented on src/pretix/control/templates/pretixcontrol/event/mail.html:
```diff
@@ -13,106 +14,13 @@
+                {% include "pretixcontrol/event/mail_settings_fragment.html" with pid="resend_link" title="Resend link" items="mail_text_resend_link,mail_text_resend_all_links" %}
+                {% include "pretixcontrol/event/mail_settings_fragment.html" with pid="order_changed" title="Order changed" items="mail_text_order_changed" %}
+                {% include "pretixcontrol/event/mail_settings_fragment.html" with pid="order_expirew" title="Payment reminder" items="mail_days_order_expire_warning,mail_text_order_expire_warning" exclude="mail_days_order_expire_warning" %}
+                {% include "pretixcontrol/event/mail_settings_fragment.html" with pid="waiting_list" title="Waiting list notification" items="mail_text_waiting_list" %}
```
You surprised me with the ``|getattr:``, I was wondering how you're going to do that but well, it seems to work and I don't have a cleaner solution right now ;)

However, we need to treat the title differently so the strings show up correctly in the translation, for example:
```
{% blocktrans asvar the_title %}Placed order{% endblocktrans %}
{% include "pretixcontrol/event/mail_settings_fragment.html" with pid="order_placed" title=the_title items="mail_text_order_placed" %}
```

Also, if you look at the ``resend_link`` part, there are two different text fields that are currently both labeled with ``Text``, we probably need to pass alternative labels here. Maybe you try to just use the ``label`` attribute of the form field? it should be filled correctly.

jlwt90 replied:
Thanks! I completely forgot about the translation part.

> Also, if you look at the resend_link part, there are two different text fields that are currently both labeled with Text, we probably need to pass alternative labels here. Maybe you try to just use the label attribute of the form field? it should be filled correctly.

I'll try using the default label attribute and see if the css works well

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -399,6 +404,79 @@ def post(self, request, *args, **kwargs):
+                idx = int(matched.group('idx'))
+                if idx < len(self.request.event.settings.locales):
+                    with translation.override(self.request.event.settings.locales[idx]):
+                        msgs[k] = v.format_map(self.placeholders(preview_item))
```
Looks good on first sight

jlwt90 commented:
>Also, pleased delete the 0052_auto_20170311_0634.py file that you accidentally created when branching of, it currently causes migrations to fail if this is merged into master.

Oops! I didn't notice that I committed migration file too. Thanks!

> Note that I'll be travelling to DjangoCon Europe starting tomorrow, so over the next week I might be slower with reviewing and replying.

No problem! Enjoy DjangoCon 👍 
Yes, I'll fix the problem you mentioned and add tests 😄

jlwt90 commented:
added tests too. I'm sorry that I somehow messed up the commits when I tried to rebase with master. It's now fixed but a list of commits appears at the bottom of this pull request..

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+            'payment_info': _('Please transfer money to this bank account: BANK-ACCOUNT-0001'),
+            'url': build_absolute_uri('presale:event.index', kwargs={
+                'event': self.request.event.slug,
+                'organizer': self.request.event.organizer
```
This needs to be ``self.request.event.organizer.slug``

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+            }),
+            'invoice_name': _('Invoice for {}.').format(self.request.event.name),
+            'invoice_company': self.request.event.organizer.name,
+            'orders': 'ORDER-0001\nORDER-0002\nORDER-0003\nORDER-0004',
```
If you look at ``pretix/presale/views/user.py``, you see that this is actually structured in the format:

```
 - EVENTSLUG-ABCDE - https://link/to/order
 - EVENTSLUG-HIJKL - https://link/to/order
```

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+                'organizer': self.request.event.organizer
+            }),
+            'invoice_name': _('Invoice for {}.').format(self.request.event.name),
+            'invoice_company': self.request.event.organizer.name,
```
This should be the name and company parts of the invoice address the *user* inputed, e.g. "John Doe" and "Sample Corporation"

jlwt90 replied:
thanks for your reminder! I checked almost all the event settings and couldn't find anything in `control` namespace. I thought they were event organiser's name & company (but I couldn't differentiate name & company in this case lol)

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+    def dummy_data(self):
+        return {
+            'event': self.request.event.name,
+            'total': 100,
```
Let's use a Decimal like 42.23 here for showing a more realistic example.

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+            'event': self.request.event.name,
+            'total': 100,
+            'currency': self.request.event.currency,
+            'date': date_format(now() + timedelta(days=7), 'SHORT_DATE_FORMAT'),
```
We also need to evaluate this in some lazy way because the date format depends of the locale. Do you have a good idea?

jlwt90 replied:
I can probably separate static data with locale related data I think.
The line`with translation.override('de'):` should be able to handle the locale based date format if we specify the formats in `pretix/helpers/formats/de/formats.py` etc.

We currently have format mappings for js file only.
```
./pretix/static/jsi18n/de/djangojs.js:    "SHORT_DATE_FORMAT": "d.m.Y",
./pretix/static/jsi18n/de-informal/djangojs.js:    "SHORT_DATE_FORMAT": "d.m.Y",
./pretix/static/jsi18n/en/djangojs.js:    "SHORT_DATE_FORMAT": "Y-m-d",
```

raphaelm replied:
That javascript is a special artifact, I'm not even sure why it's there. The translation from SHORT_DATE_FORMAT to the different formats is done within Django, not on our side.

jlwt90 replied:
got it! I realised that the problem comes from `@cached_property`. Since I cached the placeholder `{date}` and `{expire_date}` as `str`, the date value didn't change accordingly.

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+            'invoice_name': _('Invoice for {}.').format(self.request.event.name),
+            'invoice_company': self.request.event.organizer.name,
+            'orders': 'ORDER-0001\nORDER-0002\nORDER-0003\nORDER-0004',
+            'expire_date': date_format(now() + timedelta(days=15), 'SHORT_DATE_FORMAT'),
```
Same as above

raphaelm commented on src/pretix/control/views/event.py:
```diff
@@ -400,6 +405,78 @@ def post(self, request, *args, **kwargs):
+            'orders': 'ORDER-0001\nORDER-0002\nORDER-0003\nORDER-0004',
+            'expire_date': date_format(now() + timedelta(days=15), 'SHORT_DATE_FORMAT'),
+            'product': _('Admission Ticket'),
+            'hours': 12,
```
We could read the actual value from the waiting list settings here.

jlwt90 replied:
Although it sounds a little bit naive, I would like to ask a question regarding this setting.
I couldn't find the corresponding key-value pair in `pretixbase_eventsetting` table for `self.request.event.settings.waiting_list_hours`. How do I know where this setting is stored?
Would you mind giving me the direction?

raphaelm replied:
It's not naive at all :) It will only be written into the ``pretixbase_eventsetting`` when it is changed for the first time. Until then, ``self.request.event.settings.waiting_list_hours`` will give the default value.

To be correct, the default value is determined in a multi-stage way. We just this week removed this concept from pretix itself and put it into a separate library. If you want to read more about it, go here: https://django-hierarkey.readthedocs.io/en/latest/concept.html

jlwt90 commented:
@raphaelm I'll be on leave for a few days.  I'll update my source code later 😄

jlwt90 commented:
>Did I already mention that I like your code a lot? 😉 I did not yet get around to do a final review of all the code and the tests but I pulled the branch and played around with it.

Thanks! 😃 

>I found a few misunderstandings about the placeholders. They are none of your fault, its just that we documented them very poorly so far. That's also on my todo list to change…

Yes... I was actually quite confused with the meaning of some placeholders. Thank you for your clarification.

>Also, apparently my assumptions about the languages were not all correct. I have an event with en and de_Informal activated and the fields are named mail_text_resend_all_links_0 and mail_text_resend_all_links_2. Therefore, your checks don't work. If you see an easy solution to this, go ahead, but otherwise I'll deal with this mess ;)

My bad 😞  I didn't test with `en` and `de-informal`. I added the following lines to create a simple mapping for locale indices and name.
```
@cached_property
    def supported_locale(self):
        locales = {}
        for idx, val in enumerate(settings.LANGUAGES):
            if val[0] in self.request.event.settings.locales:
                locales[str(idx)] = val[0]
        return locales
```
Now we can check whether the locale is valid for this event but then I realised that my `<pre>` block generation didn't work coz the logic relied on `for loop counter`. Therefore, I changed the JSON with this format:
```
{
    'item': 'mail_text_order_free',
    'msgs': {
        'en': 'xxx',
        'de-informal': 'xxx'
    }
}
```
and some minor changes in js & html.
I guess the fix is kinda simple! 🔨 🔨  Is there any concerning point in this part? Feel free to let me know.

raphaelm commented:
Everything you say in your comment sounds perfectly fine. I'll try to find time later today or tomorrow for a final review! :)

raphaelm commented:
Hey @jlwt90,

I played around with it for a while now and everything works like a charm :sparkles: 

You did an amazing job here. This is a very good patch on so many levels, and I honestly believe I could not have done this any better, instead, it would probably would have turned out worse if I had done it. Thank you so much!

In case you are interested in taking on another issue, go have a look around if something fits your taste, otherwise I have a few suggestions: #297 and and #54 would be interesting django-y things that aren't too large. Finishing #102 could be a rather quick fix for someone as experienced as you seem to be. On the frontend side, there is the huge task of #100 where I could take some help, although I wouldn't expect from anyone to complete it, rather to take it small steps forward.

All the best,
Raphael

jlwt90 commented:
>You did an amazing job here. This is a very good patch on so many levels, and I honestly believe I could not have done this any better, instead, it would probably would have turned out worse if I had done it. Thank you so much!

Thanks! I won't be able to finish this feature without your great help 😄 
#297 looks interesting to me but I don't have any experience on android application (no problem with java coz I mostly work on java applications in my work). Do you think the task is suitable for me?
